### PR TITLE
SketchClass

### PR DESCRIPTION
**Pull request by AliHaider93**
In this pull request I will prepare the development of the beam elements with integrable cross sections

### CODE REVIEW

File: src/conv/k-g/sketch.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    if (sectionType == "SECTION_01") {
+	unsigned long vert_count = 12;
+	unsigned long seg_count  = 12;
+	line_seg* lsg;
+
+	point_t V;
+	VMOVE(V, node1);
+	vect_t r, t, s;
+	vect_t n2n3;
+	r[X] = node2[X] - node1[X];
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: VSUB2() macro?

File: src/conv/k-g/sketch.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	r[Y] = node2[Y] - node1[Y];
+	r[Z] = node2[Z] - node1[Z];
+	
+	n2n3[X] = node3[X] - node2[X];
+	n2n3[Y] = node3[Y] - node2[Y];
+	n2n3[Z] = node3[Z] - node2[Z];
+
+	t[X] = r[Y] * n2n3[Z] - r[Z] * n2n3[Y];
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: VCROSS() macro?

File: src/conv/k-g/sketch.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	t[Z] = r[X] * n2n3[Y] - r[Y] * n2n3[X];
+
+	s[X] = r[Y] * t[Z] - r[Z] * t[Y];
+	s[Y] = r[Z] * t[X] - r[X] * t[Z];
+	s[Z] = r[X] * t[Y] - r[Y] * t[X];
+
+	double normt = sqrt(t[X] * t[X] + t[Y] * t[Y] + t[Z] * t[Z]);
+	double norms = sqrt(s[X] * s[X] + s[Y] * s[Y] + s[Z] * s[Z]);
+
+	t[X] = t[X] / normt;
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: VUNITIZE() macro?

File: src/conv/k-g/sketch.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+(
+    std::string sketchName,
+    std::string sectionType,
+    const point_t& node1,
+    const point_t& node2,
+    const point_t& node3,
+    std::vector<double> D
+) {
+    if (sectionType == "SECTION_01") {
+	unsigned long vert_count = 12;
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: `const size_t`?

File: src/conv/k-g/sketch.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+rt_sketch_internal* Sketch::creatSketch
+(
+    std::string sketchName,
+    std::string sectionType,
+    const point_t& node1,
+    const point_t& node2,
+    const point_t& node3,
+    std::vector<double> D
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: `const std::vector<double>& D`?

File: src/conv/k-g/sketch.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+void Sketch::setName(const char* value)
+{
+    name = value;
+}
+
+rt_sketch_internal* Sketch::creatSketch
+(
+    std::string sketchName,
```
Comment (drossberg) [inline src/conv/k-g/sketch.cpp]: `const std::string& sketchName`? But, it is unused anyway.

File: src/conv/k-g/sketch.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+
+// Local Variables:
+// tab-width: 8
+// mode: C++
+// c-basic-offset: 4
+// indent-tabs-mode: t
+// c-file-style: "stroustrup"
+// End:
+// ex: shiftwidth=4 tabstop=8
```
Comment (drossberg) [inline src/conv/k-g/sketch.h]: The new-line at the file end is missing.

File: src/conv/k-g/sketch.cpp (hunk 1 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.cpp]: I wanted to ask if there are some macros for these kind of operations. Will modify

File: src/conv/k-g/sketch.cpp (hunk 2 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.cpp]: Also

File: src/conv/k-g/sketch.cpp (hunk 3 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.cpp]: Also

File: src/conv/k-g/sketch.cpp (hunk 4 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.cpp]: I will check

File: src/conv/k-g/sketch.cpp (hunk 6 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.cpp]: yeah I think I left it by mistake, I will check

File: src/conv/k-g/sketch.h (hunk 7 above)
Comment (AliHaider93) [reply src/conv/k-g/sketch.h]: I don't know why this alwys happens when I copy these comments :)

File: src/conv/k-g/extrude.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+) {
+    if (value != nullptr)
+	name = value;
+    else
+	name = "";
+}
+
+const char* Extrude::getName(void) const
+{
+    return nullptr;
```
Comment (drossberg) [inline src/conv/k-g/extrude.cpp]: That's an unexpected return value. Same for Sketch.
However, getName() seems to be never used from any primitive. They could all be removed?

Comment (AliHaider93) [reply src/conv/k-g/extrude.cpp]: Yes, I think I just added the functions and forgot to implement them. also you are right these functions are not uesed, I will delet them

File: src/conv/k-g/k-g.cpp [hunk 9]
```diff
@@ -162,7 +165,72 @@ int main
...
+
+					point3[X] = kData.nodes[n3].x * factor;
+					point3[Y] = kData.nodes[n3].y * factor;
+					point3[Z] = kData.nodes[n3].z * factor;
+
+					int D1 = beamSection.D[0];
+					int D2 = beamSection.D[1];
+					int D3 = beamSection.D[2];
+					int D4 = beamSection.D[3];
+
```
Comment (drossberg) [inline src/conv/k-g/k-g.cpp]: I get the following compilation errors here:
```
[ 79%] Building CXX object src/conv/k-g/CMakeFiles/k-g.dir/k-g.cpp.o
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/k-g.cpp: In function ‘int main(int, char**)’:
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/k-g.cpp:191:45: error: unused variable ‘D1’ [-Werror=unused-variable]
  191 |                                         int D1 = beamSection.D[0];
      |                                             ^~
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/k-g.cpp:192:45: error: unused variable ‘D2’ [-Werror=unused-variable]
  192 |                                         int D2 = beamSection.D[1];
      |                                             ^~
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/k-g.cpp:193:45: error: unused variable ‘D3’ [-Werror=unused-variable]
  193 |                                         int D3 = beamSection.D[2];
      |                                             ^~
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/k-g.cpp:194:45: error: unused variable ‘D4’ [-Werror=unused-variable]
  194 |                                         int D4 = beamSection.D[3];
      |                                             ^~
cc1plus: all warnings being treated as errors
```

Comment (AliHaider93) [reply src/conv/k-g/k-g.cpp]: yes, we don't need those, I will delete them

File: src/conv/k-g/extrude.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+) {
+    if (value != nullptr)
+	name = value;
+    else
+	name = "";
+}
+
+
+void Extrude::extrudeSection(std::string sectionName,const point_t& V, vect_t h, vect_t u_vec, vect_t v_vec, rt_sketch_internal* skt)
+{
```
Comment (drossberg) [inline src/conv/k-g/extrude.cpp:50]: I get the following compilation errors here:
```[ 79%] Building CXX object src/conv/k-g/CMakeFiles/k-g.dir/extrude.cpp.o
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/extrude.cpp: In member function ‘void Extrude::extrudeSection(std::string, const fastf_t (&)[3], fastf_t*, fastf_t*, fastf_t*, rt_sketch_internal*)’:
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/extrude.cpp:49:130: error: unused parameter ‘skt’ [-Werror=unused-parameter]
   49 | on(std::string sectionName,const point_t& V, vect_t h, vect_t u_vec, vect_t v_vec, rt_sketch_internal* skt)
      |                                                                                    ~~~~~~~~~~~~~~~~~~~~^~~
```

File: src/conv/k-g/extrude.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+std::string Extrude::write(rt_wdb* wdbp)
+{
+    std::string ret;
+
+    rt_extrude_internal* extrude_wdb;
+    BU_GET(extrude_wdb, rt_extrude_internal);
+    extrude_wdb->magic = RT_EXTRUDE_INTERNAL_MAGIC;
+    extrude_wdb = m_extrude;
+
+    if (extrude_wdb->sketch_name != "") {
```
Comment (drossberg) [inline src/conv/k-g/extrude.cpp]: I get the following compilation errors here:
```
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/extrude.cpp: In member function ‘std::string Extrude::write(rt_wdb*)’:
/home/rossberg/Devel/BRL-CAD/brlcad.devel_k-g/src/conv/k-g/extrude.cpp:71:34: error: comparison with string literal results in unspecified behavior [-Werror=address]
   71 |     if (extrude_wdb->sketch_name != "") {
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```
Using **strcmp()**?

Comment (drossberg) [reply src/conv/k-g/extrude.cpp]: Or
```
(extrude_wdb->sketch_name != nullptr) && (strlen(extrude_wdb->sketch_name) > 0)
```
?

File: src/conv/k-g/extrude.cpp (hunk 10 above)
Comment (AliHaider93) [reply src/conv/k-g/extrude.cpp:50]: yes, I though we might need it for the extrude. I have to remove it

File: src/conv/k-g/extrude.cpp (hunk 11 above)
Comment (AliHaider93) [reply src/conv/k-g/extrude.cpp]: second choice is better

**Pull request merged**